vercel/next.js
# feat(middleware): issues warnings when using node.js global APIs in middleware

## What's in there?

Middleware, since they are ran on edge runtime, do not have access to Node.js globals. While #36434 addressed cases where a middleware would import a node.js module, this PR issues warning when trying to access unsupported globals, such as:
`clearImmediate()`, `process.cwd()`, `Event` or `TextDecoderStream` (full list in `packages/next/shared/lib/constants.ts`).

The new behavior is:
- a warning at build time. Failing the build would be too strict: it's possible for the final bundle to include unsupported calls that would never be reached at run time (despite tree-shaking).
- a warning on the first access at run time (useful in development)
- `process.env` is always allowed
- other attributes/methods of `process` are not supported, but could be overridden

## How to test?

First, build next: `yarn workspace next release`

Then, using integration tests:
1. `yarn jest test/integration/middleware-with-node.js-apis` (checks both build and run time behavior)
1. `yarn jest test/integration/middleware-overrides-node.js-api` (ability to override `process.XYZ` and use `process.env` with no warning)

Finally, as an example:
1. create a middleware file in `examples/with-node-api/middleware.js`, with this content:
   ```js
   import { NextResponse } from 'next/server'
   
   export default function middleware() {
     console.log(`this should work:`, process.env, __dirname, __filename)
     console.log(`this should not`, setImmediate(() => {}))
     return NextResponse.next()
   }
   ```
1. creates a plain pages in  `examples/with-node-api/pages/index.js`, with this content:
   ```js
   export default function Home() {
     return <div>A page</div>
   }
   ```
3. run it in dev mode  `yarn next dev examples/with-node-api/` and browse to https://localhost:3000:
   > it should print `this should work: { NEXT_RUNTIME: 'edge' }`, issue a warning about using `setImmediate()`, and fail because `setImmediate()` is not detined.
4. build it: `yarn next build examples/with-node-api/`:
   > it should build, and issue a warning about using `setImmediate()`,

## Notes to reviewers

- `build/webpack/plugins/middleware-plugin.ts` does the build-time validation.
- `server/web/sandbox/context.ts` does the run-time behavior.

To determine unsupported APIs, I compared the list from [node.js docs](https://nodejs.org/docs/latest-v17.x/api/globals.html) to the our [edge-runtime](https://nextjs.org/docs/api-reference/edge-runtime).
Please note that `__dirname` and `__filename` are still allowed because webpack uses them.

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [x] Related issues linked using `fixes #number`
- [ ] Telemetry added. In case of a feature if it's used or not.
- [x] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [x] Make sure the linting passes by running `yarn lint`
